### PR TITLE
Python: Bottle Framework Support

### PR DESCRIPTION
**Pull request by Kwstubbs**
Add basic Bottle support

### CODE REVIEW

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Bottle.qll:103]: This code looks like it will work for both reads and writes to `bottle.response.headers`? 

might want to just do the same as we do for flask/werkzeug: https://github.com/github/codeql/blob/d4a072818f451a281d9faddaf44fab3264d457fc/python/ql/lib/semmle/python/frameworks/Werkzeug.qll#L216-L239

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Thanks :tada: 

Overall the code looks OK to me, but to accept this PR I need you to write some tests :blush: (good start by copying over the ConceptsTest, but we also need some code using bottle to show that the modeling works as intended) See [flask tests](https://github.com/github/codeql/tree/main/python/ql/test/library-tests/frameworks/flask) as an inspiration.

**Comment by Kwstubbs** [comment]
@RasmusWL I've added more support for Bottle routes and I've added tests.

**Comment by yoff** [comment]
@RasmusWL is travelling at the moment, so I will take this over for now.
I am curious about the switch away from API nodes in https://github.com/github/codeql/pull/17370/commits/5d12f7bd30a87900e818f230189d6c6ee88b7d59, was it necessary to get the tests pass?
It looks like you based this off of other models. This is absolutely fine, but let us double check that we are actually modelling the bottle framework.

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [inline python/ql/lib/semmle/python/frameworks/Bottle.qll:36]: I think `app` (like `appstack`) is actually a list that you would have to `pop`, see https://bottlepy.org/docs/dev/api.html#bottle.app and https://bottlepy.org/docs/dev/api.html#bottle.AppStack.pop
```suggestion
      API::Node app() { result = bottle().getMember("Bottle").getReturn() }
```

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [inline python/ql/lib/semmle/python/frameworks/Bottle.qll:30]: This seems to only be used in `instance`, perhaps just inline it?

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [inline python/ql/lib/semmle/python/frameworks/Bottle.qll:56]: It seems the list of calls are the same in all situations by design of the bottle framework, so we could have a string predicate for it. That would ensure that we do not forget one function in one of the cases.
Alternatively, you could factor out the `API::Node`, like `exists(API::Node app | app in [App::instance(), App::app(), bottle() | this = app..getMember(["route", "get", "post", "put", "delete", "patch"])
                  .getACall() )`

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [inline python/ql/lib/semmle/python/frameworks/Bottle.qll:75]: Shoudl we include `bottle.Response.getReturn()` (with capital R) here? See https://bottlepy.org/docs/dev/api.html#the-response-object.

File: python/ql/test/library-tests/frameworks/bottle/taint_test.py
Comment (yoff) [inline python/ql/test/library-tests/frameworks/bottle/taint_test.py:5]: I am not sure this works, where did you find this? I think the idiomatic example is here: https://bottlepy.org/docs/dev/tutorial.html#the-application-object

**Review by yoff — CHANGES_REQUESTED** [review]
The structure of both the code and the tests are good. I am a little worried that the modelling is not completely aligned with the framework; all these frameworks have subtle differences. A good way to align with the framework could be to write several examples, taken from tutorials and the like, in the test files and then align the modelling to those (padding out the set of function names and argument names by looking at the documentation).

File: python/ql/test/library-tests/frameworks/bottle/taint_test.py
Comment (Kwstubbs) [reply python/ql/test/library-tests/frameworks/bottle/taint_test.py:5]: You can find it in some of the [documentation](https://bottle.readthedocs.io/en/stable/api.html#bottle.app). There are also a lot of examples on Github search.

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (Kwstubbs) [reply python/ql/lib/semmle/python/frameworks/Bottle.qll:36]: The documentation says "A stack-like list. Calling it returns the head of the stack." so this should be fine.

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (Kwstubbs) [reply python/ql/lib/semmle/python/frameworks/Bottle.qll:75]: I'm modeling https://bottlepy.org/docs/dev/api.html#bottle.response which is global and not a function. Don't think its necessary.

**Comment by Kwstubbs** [comment]
@yoff Hi, please let me know what you think of the changes. I'm trying to model  every possible usage of bottle, so even though the documentation might not mention something, if I find it is used a lot in Github I have included them. Obviously it must run on the latest, currently v12.25 as LTS. Also, I'm not sure where I moved away from API nodes in [5d12f7bd3](https://github.com/github/codeql/commit/5d12f7bd30a87900e818f230189d6c6ee88b7d59). Everything I have modeled here I have created a codebase/codeql database to ensure validity.

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Bottle.qll:33]: Indeed, that looks ok :-)

File: python/ql/lib/semmle/python/frameworks/Bottle.qll
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Bottle.qll:72]: I think it is an additional may to obtain/modify responses. It may only see modest use ([partial ci search](https://github.com/search?type=code&q=%2F%28%3F-i%29bottle%5C.Response%2F+language%3APython)), but it should be easy enough for us to recognise it?

File: python/ql/test/library-tests/frameworks/bottle/taint_test.py
Comment (yoff) [reply python/ql/test/library-tests/frameworks/bottle/taint_test.py:5]: Ok, that is great to hear :-) Should we also test that we support the simple example from the tutorial, though?

**Comment by yoff** [comment]
> Also, I'm not sure where I moved away from API nodes in [5d12f7bd3](https://github.com/github/codeql/commit/5d12f7bd30a87900e818f230189d6c6ee88b7d59).

Sorry, that was a somewhat vague reference to a large diff. I meant the fields `name` and `value` inside the class `HeaderWriteSubscript` in the `Headers` module (see [here](https://github.com/github/codeql/commit/5d12f7bd30a87900e818f230189d6c6ee88b7d59#diff-de17e498e80d1fbbcf311be51b22b58c6dc32ab3f10e75c23fb0f8a493059b5eL93-R163)).

**Review by yoff — CHANGES_REQUESTED** [review]
I am requesting a small simplification, but I am feeling much better about this now; thanks for explaining your testing :-)

File: python/ql/test/library-tests/frameworks/bottle/taint_test.py
Comment (Kwstubbs) [reply python/ql/test/library-tests/frameworks/bottle/taint_test.py:5]: Added

**Comment by Kwstubbs** [comment]
@yoff I was just copying from [Werkzeug.qll](https://github.com/github/codeql/blob/d4a072818f451a281d9faddaf44fab3264d457fc/python/ql/lib/semmle/python/frameworks/Werkzeug.qll#L216-L239) as recommended by Rasmus on his first pass. The API nodes was working, as I was using these models in my research before PRing. ✅

**Review by yoff — APPROVED** [review]
LGTM

**Comment by yoff** [comment]
> @yoff I was just copying from [Werkzeug.qll](https://github.com/github/codeql/blob/d4a072818f451a281d9faddaf44fab3264d457fc/python/ql/lib/semmle/python/frameworks/Werkzeug.qll#L216-L239) as recommended by Rasmus on his first pass. The API nodes was working, as I was using these models in my research before PRing. ✅

In that case, it would be nicer to use the API nodes. They are cleaner and might match in a few more cases. I will not require it, though, it can be a clean-up of both situations later.

**Comment by Kwstubbs** [comment]
@yoff just curious about the status of this PR. Are we just waiting on approval from @RasmusWL?

**Review by yoff — APPROVED** [review]
LGTM

**Pull request merged**
